pydantic/function_model.py

```diff
@@ -0,0 +1,442 @@
+"""Provides FunctionModel, a model for validating function arguments."""
+
+from collections import OrderedDict
+from copy import deepcopy
+from inspect import Parameter, Signature, signature
+from types import NoneType
+from typing import Any, Callable, List, Optional, Tuple, Type, TypeVar, Union
+
+from pydantic_core import InitErrorDetails, PydanticCustomError, PydanticKnownError, ValidationError
+
+from pydantic import BaseModel, ConfigDict, TypeAdapter, create_model, errors
+
+
+class FunctionModel:
+    """A class for validating function arguments and calling the function upon the validated arguments."""
+
+    function: Callable
+    _model: Type[BaseModel]
+    _parsed: Optional[BaseModel]
+    _return: TypeAdapter
+    _parameters: OrderedDict
+    _keyword_total: int
+    _positional_total: int
+
+    T = TypeVar('T')
+
+    def __init__(self, function: Callable) -> None:
+        """Create a new model based on the provided function.
+
+        Args:
+            function: The function that the model should encapsulate.
+        """
+        self.function = function
+        self._parameters = OrderedDict()
+        self._parsed = None
+        self._keyword_total = 0
+        self._positional_total = 0
+
+        # Gather all parameters from the function
+        params = OrderedDict()  # In order, as params[<name>] = (type, default)
+        has_args = False
+        has_kwargs = False
+        sig = signature(function)
+        for name, param in sig.parameters.items():
+            if name == 'args' and param.kind != Parameter.VAR_POSITIONAL:
+                raise ValueError('The name "args" can only be used for variables in the case of *args.')
+            elif name == 'kwargs' and param.kind != Parameter.VAR_KEYWORD:
+                raise ValueError('The name "kwargs" can only be used for variables in the case of **kwargs.')
+            elif name.startswith('_'):
+                raise _internal_key_error(name)
+
+            if param.kind == Parameter.VAR_POSITIONAL:
+                has_args = True
+            elif param.kind == Parameter.VAR_KEYWORD:
+                has_kwargs = True
+            else:
+                is_positional = param.kind == Parameter.POSITIONAL_ONLY or param.kind == Parameter.POSITIONAL_OR_KEYWORD
+                is_keyword = param.kind == Parameter.KEYWORD_ONLY or param.kind == Parameter.POSITIONAL_OR_KEYWORD
+
+                if is_positional:
+                    self._positional_total += 1
+                if is_keyword:
+                    self._keyword_total += 1
+
+                params[name] = (
+                    get_annotation_type(param.annotation),
+                    param.default if param.default != Parameter.empty else ...,
```

Comment:
I think this won't work if the default value is `...`.

Reply:
I got the impression somewhere (possibly [here](https://github.com/pydantic/pydantic/blob/b46010bf4a73d439e3faf98e6e2e16e2faf7e6e2/pydantic/fields.py#L634), I'm not sure) that ... was the default value of field defaults. From looking further at that class, it seems they use PydanticUndefined, which I have changed it to and works just as well with the current tests. I can't find any documentation as to what exactly I _should_ use here, but hopefully if there is an issue then it will get pointed out once we have an upstream PR - I'll make sure to ask explicitly when the time comes so it's not missed.